File: 395_Longest_Substring_with_At_Least_K_Repeating_Characters/395_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
#include <sstream>
#include <set>
#include <omp.h>
using namespace std;

class Solution {
public:
    int longestSubstring(string s, int k) {
        int n = s.size();
        unordered_map<char, int> map;
        for (char c : s) map[c]++;
        vector<int> split;
        for (int i = 0; i < n; i++) {
            if (map[s[i]] < k) split.emplace_back(i);
        }
        if (split.empty()) return n;
        split.emplace_back(n);
        int res = 0, left = 0;
        for (int sp : split) {
            int len = sp - left;
            if (len > res)
                return max(res, longestSubstring(s.substr(left, len), k));
            left = sp + 1;
        }
        return res;
    }
};

int main() {
    Solution s;
    cout << s.longestSubstring("ababbc", 2) << endl;
}

File: 395_Longest_Substring_with_At_Least_K_Repeating_Characters/395_1.py
from typing import List, Dict

class Solution(object):
    def longestSubstring(self, s, k):
        if len(s) < k:
            return 0
        c = min(set(s), key=s.count)
        if s.count(c) >= k:
            return len(s)
        return max(self.longestSubstring(t, k) for t in s.split(c))


if __name__ == "__main__":
    print(Solution().longestSubstring("ababbc", 2))
